File: src/main/scala/com/vc/medium/L503.scala
package com.vc.medium

object L503 {
  def nextGreaterElements(arr: Array[Int]): Array[Int] = {
    val n = arr.length
    val next = Array.fill(n)(-1)
    import scala.collection.mutable
    val st = new mutable.Stack[Int]()
    (0 until 2 * n).foreach(i => {
      val num = arr(i % n)
      while(st.nonEmpty && arr(st.head) < num) {
        next(st.pop) = num
      }
      if(i < n) st.push(i)
    })
    next
  }
}

File: src/main/scala/com/vc/medium/L503.java
package com.vc.medium;

import java.util.*;

class L503 {
    public int[] nextGreaterElements(int[] arr) {
        int n = arr.length;
        int[] res = new int[n];
        Arrays.fill(res, -1);
        Stack<Integer> st = new Stack<>();
        solve(arr, res, st);
        solve(arr, res, st);
        return res;
    }

    private void solve(int[] arr, int res[], Stack<Integer> st) {
        for(int i = 0; i < arr.length; i++) {
            while(!st.isEmpty() && arr[st.peek()] < arr[i]) res[st.pop()] = arr[i];
            st.push(i);
        }
    }
}